File: apps/enrichment-session/src/session.ts
import type { RedisClient } from '@org/redis_client';
import type { SessionState } from './types.js';

const RAGE_WINDOW_MS = 30_000;
const RAGE_THRESHOLD = 3;
const MAX_RAGE_TIMESTAMPS = 10;

/** Context fields written once per session for downstream consumers (decision-engine stale scanner). */
export interface SessionContext {
  storeId: number;
  customerId: number | null;
  distinctId: string;
  anon?: string;
  uid?: string;
  email?: string;
  emailConsent: boolean;
  smsConsent: boolean;
}

export class SessionService {
  constructor(
    private readonly redis: RedisClient,
    private readonly ttlSeconds: number,
  ) {}

  async updateSession(
    sessionId: string,
    eventType: string,
    cartValueDelta: number,
    nowMs: number,
  ): Promise<SessionState> {
    const raw = this.redis.getRedis();
    const sessionKey = `session:${sessionId}`;
    const rageTsKey = `session:${sessionId}:rage_ts`;

    const isRageClick = eventType === 'rage_click';

    // Pipeline: update hash fields atomically
    const pipeline = raw.pipeline();
    if (cartValueDelta !== 0) {
      pipeline.hincrbyfloat(sessionKey, 'cart_value', cartValueDelta);
    }
    if (isRageClick) {
      pipeline.hincrby(sessionKey, 'rage_click_count', 1);
    }
    pipeline.hset(sessionKey, 'last_activity', nowMs);
    pipeline.expire(sessionKey, this.ttlSeconds);

    // Track this session in the active:sessions sorted set (used by stale scanner)
    pipeline.zadd('active:sessions', nowMs, sessionId);

    if (isRageClick) {
      pipeline.lpush(rageTsKey, nowMs);
      pipeline.ltrim(rageTsKey, 0, MAX_RAGE_TIMESTAMPS - 1);
      pipeline.expire(rageTsKey, this.ttlSeconds);
    }

    await pipeline.exec();

    // Read back current state — hgetall returns null for a missing key
    const [hashData, rageTimestamps] = await Promise.all([
      raw.hgetall(sessionKey),
      raw.lrange(rageTsKey, 0, -1),
    ]);

    const cartValue = parseFloat(hashData?.['cart_value'] ?? '0');
    const rageClickCount = parseInt(hashData?.['rage_click_count'] ?? '0', 10);
    const lastActivity = parseInt(hashData?.['last_activity'] ?? String(nowMs), 10);

    // Sliding window: count rage clicks in last RAGE_WINDOW_MS
    const recentRageClicks = rageTimestamps.filter(
      (ts) => nowMs - parseInt(ts, 10) <= RAGE_WINDOW_MS,
    ).length;
    const isFrustrated = recentRageClicks >= RAGE_THRESHOLD;

    // Persist is_frustrated so the decision-engine can read it without recomputing
    await raw.hset(sessionKey, 'is_frustrated', isFrustrated ? '1' : '0');

    return { cartValue, rageClickCount, lastActivity, isFrustrated };
  }

  /**
   * Writes stable identity context fields to the session hash.
   * Called by the Enricher after resolving customer data so the decision-engine's
   * scheduler worker and stale scanner can reconstruct a full event without Kafka.
   * Uses HSET (overwrites) so the latest resolved customer_id is always stored.
   */
  async setContext(sessionId: string, ctx: SessionContext): Promise<void> {
    const raw = this.redis.getRedis();
    const sessionKey = `session:${sessionId}`;
    await raw.hset(sessionKey, {
      store_id: String(ctx.storeId),
      customer_id: ctx.customerId !== null ? String(ctx.customerId) : '',
      distinct_id: ctx.distinctId,
      anon: ctx.anon ?? ctx.distinctId,
      uid: ctx.uid ?? '',
      email: ctx.email ?? '',
      email_consent: ctx.emailConsent ? '1' : '0',
      sms_consent: ctx.smsConsent ? '1' : '0',
    });
  }
}

File: apps/enrichment-session/src/session.rs
use std::sync::Arc;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum SessionError {
    #[error("Redis error: {0}")]
    RedisError(String),
}

#[derive(Debug, Clone)]
pub struct SessionContext {
    pub store_id: i32,
    pub customer_id: Option<i32>,
    pub distinct_id: String,
    /// Anonymous visitor ID alias — falls back to distinct_id if not set.
    pub anon: Option<String>,
    /// Authenticated user ID (empty string when not authenticated).
    pub uid: Option<String>,
    pub email: Option<String>,
    pub email_consent: bool,
    pub sms_consent: bool,
}

pub struct SessionService {
    redis: Arc<redis::Client>,
}

impl SessionService {
    pub fn new(redis: Arc<redis::Client>, _ttl_seconds: u64) -> Self {
        Self { redis }
    }

    /// Write identity context to session hash.
    pub async fn set_context(&self, session_id: &str, ctx: SessionContext) -> Result<(), SessionError> {
        let mut con = self.redis.get_multiplexed_async_connection().await
            .map_err(|e| SessionError::RedisError(e.to_string()))?;

        let session_key = format!("session:{}", session_id);
        let anon = ctx.anon.as_deref().unwrap_or(ctx.distinct_id.as_str()).to_string();
        let uid = ctx.uid.as_deref().unwrap_or("").to_string();
        let _: () = redis::cmd("HSET")
            .arg(&session_key)
            .arg("store_id")
            .arg(ctx.store_id.to_string())
            .arg("customer_id")
            .arg(ctx.customer_id.map(|id| id.to_string()).unwrap_or_default())
            .arg("distinct_id")
            .arg(&ctx.distinct_id)
            .arg("anon")
            .arg(anon)
            .arg("uid")
            .arg(uid)
            .arg("email")
            .arg(ctx.email.unwrap_or_default())
            .arg("email_consent")
            .arg(if ctx.email_consent { "1" } else { "0" })
            .arg("sms_consent")
            .arg(if ctx.sms_consent { "1" } else { "0" })
            .query_async::<_, ()>(&mut con)
            .await
            .map_err(|e| SessionError::RedisError(e.to_string()))?;

        Ok(())
    }
}
